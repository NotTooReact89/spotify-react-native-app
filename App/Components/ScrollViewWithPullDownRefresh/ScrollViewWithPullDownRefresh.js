// flow
import * as React from 'react';
import { RefreshControl, View, Image } from 'react-native';
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import Banner from '../Banner';
import SyncActions, {
  isSyncing,
  getSyncError,
  getSyncErrorId,
  isOnScreenMessageVisible,
} from '../../Redux/SyncRedux';
import type { Error } from '../../Redux/types';
import { TestProps, Analytics, ApiUtils } from '../../Lib';
import { isConnectedToInternet } from '../../Redux/StartupRedux';
import { getEmployeeId } from '../../Redux/Authentication/UserProfileRedux';
import { Colors } from '../../Styles';
import Text from '../Text';
import Styles from './ScrollViewWithPullDownRefreshStyles';

type Props = {
  background: any,
  contentContainerStyle?: any, // $FlowFixMe,
  onKeyboardDidShow?: () => void,
  onKeyboardDidHide?: () => void,
  showExtraBanner?: boolean,
  banner?: React.Node,
  children: React.Node,
  screenName: string,
  isConnectedToInternet: boolean,
  startSync: () => void,
  isSyncing: boolean,
  syncError: ?Error,
  syncErrorId: ?string,
  showAppBanner: boolean,
  hideOnScreenMessage: () => void,
};

export class ScrollViewWithPullDownRefresh extends React.Component<Props, *> {
  props: Props;

  static defaultProps = {
    contentContainerStyle: {},
    onKeyboardDidShow: () => {},
    onKeyboardDidHide: () => {},
    showExtraBanner: false,
    banner: null,
  };

  onRefresh = () => {
    this.props.startSync();
  };

  render() {
    return (
      <View style={Styles.container}>
        {this.props.syncError && this.props.showAppBanner ? (
          <Banner
            text={ApiUtils.convertApiErrorToText(this.props.syncError)}
            onPress={this.props.hideOnScreenMessage}
          />
        ) : null}
        {this.props.showExtraBanner ? this.props.banner : null}

        <KeyboardAwareScrollView
          contentContainerStyle={this.props.contentContainerStyle}
          keyboardShouldPersistTaps="handled"
          onKeyboardDidShow={this.props.onKeyboardDidShow}
          onKeyboardDidHide={this.props.onKeyboardDidHide}
          refreshControl={
            <RefreshControl
              refreshing={this.props.isSyncing}
              onRefresh={this.onRefresh}
              tintColor={Colors.base.textInverse}
              style={{ backgroundColor: 'black' }}
            />
          }
        >
          {this.props.children}
        </KeyboardAwareScrollView>
      </View>
    );
  }
}

const mapStateToProps = state => ({
  isSyncing: isSyncing(state),
  isConnectedToInternet: isConnectedToInternet(state),
  syncError: getSyncError(state),
  showAppBanner: isOnScreenMessageVisible(state),
});

export const mapDispatchToProps = (dispatch: Dispatch<*>) => ({
  startSync: () => {
    dispatch(SyncActions.startSync());
  },
  hideOnScreenMessage: () => {
    dispatch(SyncActions.hideOnScreenMessage());
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ScrollViewWithPullDownRefresh);
