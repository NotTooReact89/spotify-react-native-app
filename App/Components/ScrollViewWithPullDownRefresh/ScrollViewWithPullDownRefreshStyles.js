// @flow

import { StyleSheet } from 'react-native';
import { Colors, Spaces, BaseStyles } from '../../Styles';

export default StyleSheet.create({
  ...BaseStyles,

  container: {
    position: 'relative',
    height: '100%',
    backgroundColor: Colors.base.text,
  },
});
