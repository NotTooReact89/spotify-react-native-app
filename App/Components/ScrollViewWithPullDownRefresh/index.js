// @flow

export { default } from './ScrollViewWithPullDownRefresh'
