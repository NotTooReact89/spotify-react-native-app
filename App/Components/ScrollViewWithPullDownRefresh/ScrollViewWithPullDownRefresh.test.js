// @flow

import React from 'react'
import { View } from 'react-native'
import renderer from 'react-test-renderer'
import { ScrollViewWithPullDownRefresh } from './ScrollViewWithPullDownRefresh'

const props = {
  background: 'morningLight',
  hideOnScreenMessage: jest.fn(),
  isConnectedToInternet: true,
  screenName: 'testScreen',
  startSync: jest.fn(),
  userId: '12345',
}

describe('render notification banner when sync fails', () => {
  it('with error id', () => {
    const tree = renderer.create(
      <ScrollViewWithPullDownRefresh
        {...props}
        syncErrorId="23456"
        syncError="genericError"
        showAppBanner
        isSyncing={false}
      >
        <View />
      </ScrollViewWithPullDownRefresh>
    )
    expect(tree).toMatchSnapshot()
  })

  it('without error id', () => {
    const tree = renderer.create(
      <ScrollViewWithPullDownRefresh
        {...props}
        syncErrorId=""
        syncError="genericError"
        showAppBanner
        isSyncing={false}
      >
        <View />
      </ScrollViewWithPullDownRefresh>
    )
    expect(tree).toMatchSnapshot()
  })
})

it('should render scroll view with pull down to refresh', () => {
  const tree = renderer.create(
    <ScrollViewWithPullDownRefresh
      {...props}
      syncErrorId=""
      syncError={null}
      showAppBanner={false}
      isSyncing={false}
    >
      <View />
    </ScrollViewWithPullDownRefresh>
  )
  expect(tree).toMatchSnapshot()
})

it('should render loading icon when sync is running', () => {
  const tree = renderer.create(
    <ScrollViewWithPullDownRefresh
      {...props}
      syncErrorId=""
      syncError={null}
      showAppBanner={false}
      isSyncing
    >
      <View />
    </ScrollViewWithPullDownRefresh>
  )
  expect(tree).toMatchSnapshot()
})
