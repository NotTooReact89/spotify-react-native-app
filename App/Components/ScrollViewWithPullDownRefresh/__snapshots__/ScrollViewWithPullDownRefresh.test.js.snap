// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render notification banner when sync fails with error id 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000000",
      "height": "100%",
      "position": "relative",
    }
  }
>
  <View
    style={
      Object {
        "transform": Array [
          Object {
            "translateY": -100,
          },
        ],
      }
    }
    testID="banner_container"
  >
    <View
      style={
        Object {
          "backgroundColor": "#008392",
          "flexDirection": "row",
          "padding": 20,
          "zIndex": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "process": [Function],
            }
          }
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "resizeMode": "contain",
              },
              Object {
                "height": 24,
                "width": 24,
              },
              Object {},
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontFamily": "Lato-Light",
                "fontSize": 15,
                "lineHeight": 22.5,
              },
              Object {
                "flex": 1,
                "marginLeft": 20,
              },
            ]
          }
          testID="banner_txtMessage"
        >
          There was a problem fetching data. Please try again later.
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "paddingLeft": 40,
          }
        }
        testID="banner_btnClose"
      >
        <Image
          source={
            Object {
              "process": [Function],
            }
          }
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "resizeMode": "contain",
              },
              Object {
                "height": 24,
                "width": 24,
              },
              Object {},
            ]
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={Object {}}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onKeyboardDidHide={[Function]}
    onKeyboardDidShow={[Function]}
    onScroll={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
        style={
          Object {
            "backgroundColor": "black",
          }
        }
        tintColor="#ffffff"
      />
    }
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <RCTRefreshControl />
    <View>
      <View />
    </View>
  </RCTScrollView>
</View>
`;

exports[`render notification banner when sync fails without error id 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000000",
      "height": "100%",
      "position": "relative",
    }
  }
>
  <View
    style={
      Object {
        "transform": Array [
          Object {
            "translateY": -100,
          },
        ],
      }
    }
    testID="banner_container"
  >
    <View
      style={
        Object {
          "backgroundColor": "#008392",
          "flexDirection": "row",
          "padding": 20,
          "zIndex": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "process": [Function],
            }
          }
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "resizeMode": "contain",
              },
              Object {
                "height": 24,
                "width": 24,
              },
              Object {},
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontFamily": "Lato-Light",
                "fontSize": 15,
                "lineHeight": 22.5,
              },
              Object {
                "flex": 1,
                "marginLeft": 20,
              },
            ]
          }
          testID="banner_txtMessage"
        >
          There was a problem fetching data. Please try again later.
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "paddingLeft": 40,
          }
        }
        testID="banner_btnClose"
      >
        <Image
          source={
            Object {
              "process": [Function],
            }
          }
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "resizeMode": "contain",
              },
              Object {
                "height": 24,
                "width": 24,
              },
              Object {},
            ]
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={Object {}}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onKeyboardDidHide={[Function]}
    onKeyboardDidShow={[Function]}
    onScroll={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
        style={
          Object {
            "backgroundColor": "black",
          }
        }
        tintColor="#ffffff"
      />
    }
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <RCTRefreshControl />
    <View>
      <View />
    </View>
  </RCTScrollView>
</View>
`;

exports[`should render loading icon when sync is running 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000000",
      "height": "100%",
      "position": "relative",
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={Object {}}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onKeyboardDidHide={[Function]}
    onKeyboardDidShow={[Function]}
    onScroll={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={true}
        style={
          Object {
            "backgroundColor": "black",
          }
        }
        tintColor="#ffffff"
      />
    }
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <RCTRefreshControl />
    <View>
      <View />
    </View>
  </RCTScrollView>
</View>
`;

exports[`should render scroll view with pull down to refresh 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000000",
      "height": "100%",
      "position": "relative",
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={Object {}}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onKeyboardDidHide={[Function]}
    onKeyboardDidShow={[Function]}
    onScroll={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
        style={
          Object {
            "backgroundColor": "black",
          }
        }
        tintColor="#ffffff"
      />
    }
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <RCTRefreshControl />
    <View>
      <View />
    </View>
  </RCTScrollView>
</View>
`;
